src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,381 @@
+                        cachedHeaders.Add(headerCache);
+                    else
+                        blockHeightList.Add(heightKey);
+
+                    heightKey++;
+
+                } while (heightKey <= toBlockHeight);
+
+                // Try and get items from the repository if not found in the store cache.
+                if (blockHeightList.Count > 0)
+                {
+                    var repositoryHeaders = new List<ProvenBlockHeader>();
+
+                    foreach(int blockHeight in blockHeightList)
+                    {
+                        var repositoryHeader = new ProvenBlockHeader();
+
+                        repositoryHeader = await this.provenBlockHeaderRepository.GetAsync(blockHeight).ConfigureAwait(false);
+
+                        if (repositoryHeader != null)
+                        {
+                            repositoryHeaders.Add(repositoryHeader);
+
+                            // When found also add to cache.
+                            this.Cache.AddOrUpdate(blockHeight, repositoryHeader);
+                        }
+                    }
+
+                    if (repositoryHeaders.Count > 0)
+                        cachedHeaders.AddRange(repositoryHeaders);
+                }
+            }
+
+            return cachedHeaders;
+        }
+
+        /// <inheritdoc />
+        public async Task<HashHeightPair> CurrentTipHashHeightAsync()
+        {
+            this.CurrentTipHashHeight = await this.provenBlockHeaderRepository.GetTipHashHeightAsync().ConfigureAwait(false);
+
+            return this.CurrentTipHashHeight;
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetTipAsync()
+        {
+            this.CurrentTipHashHeight = await this.CurrentTipHashHeightAsync();
+
+            return await this.GetAsync(this.CurrentTipHashHeight.Height).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void AddToPending(ProvenBlockHeader provenBlockHeader, HashHeightPair newTip)
+        {
+            this.PendingCache.AddOrUpdate(newTip.Height, provenBlockHeader, (key, value) => { return provenBlockHeader; });
+
+            this.pendingTipHashHeight = newTip;
+        }
+
+        /// <summary>Saves pending <see cref="ProvenBlockHeader"/> items to the <see cref="ProvenBlockHeaderRepository"/>.</summary>
+        private async Task SaveAsync()
+        {
+            if (this.PendingCache.Count == 0)
+            {
+                this.logger.LogTrace("(-)[PROVEN_BLOCK_HEADER_CACHE_EMPTY]");
+                return;
+            }
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
+            {
+                using (await this.lockobj.LockAsync().ConfigureAwait(false))
+                {
+                    var pendingHeaders = new List<ProvenBlockHeader>();
+
+                    for (int i = this.CurrentTipHashHeight.Height; i <= this.pendingTipHashHeight.Height; i++)
```

Comment:
What if some headers are left in pending after this? 

Why not take whole pending and save to underlying database?

Reply:
Ok, @dangershony also made this point.  I've changed.